pandas/io/common.py

```diff
@@ -128,9 +129,33 @@ def _validate_header_arg(header) -> None:
         )
 
 
-def _stringify_path(
-    filepath_or_buffer: FilePathOrBuffer[AnyStr]
-) -> FilePathOrBuffer[AnyStr]:
+# Overload *variants* for '_stringify_path'.
+# These variants give extra information to the type checker.
+# They are ignored at runtime.
+
+
+@overload
+def _stringify_path(filepath_or_buffer: Union[str, Path]) -> str:
+    ...
+
+
+@overload
+def _stringify_path(filepath_or_buffer: IO[AnyStr]) -> IO[AnyStr]:
+    ...
+
+
+# The actual *implementation* of '_stringify_path'.
```

Comment:
The comment is copy and pasted from https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading

When adding the overloads I removed the type annotations from the actual implementation to avoid duplication.

I guess this was the wrong thing to do. will add them back.